src/prowjobsscraper/equinix_usages.py

```diff
@@ -108,4 +108,68 @@ def get_project_usages(
             headers={self._EQUINIX_ENDPOINT_HEADER: self._project_token},
         ).json()["usages"]
         logger.info("%s usages retrieved successfully", len(equinix_project_usages))
-        return [EquinixUsage.parse_obj(usage) for usage in equinix_project_usages]
+        return self._process_usages(
+            [EquinixUsage.parse_obj(usage) for usage in equinix_project_usages]
+        )
+
+    def _is_usage_in_interval(self, usage: EquinixUsage) -> bool:
+        """Usage is considered to be within the time interval
+        if its complete duration falls within that interval
+        and it has reached completion.
+        """
+        return (
+            usage.end_date is not None
+            and usage.start_date >= self._start_time
+            and usage.end_date <= self._end_time
+        )
+
+    def _process_usages(cls, usages: list[EquinixUsage]) -> list[EquinixUsage]:
+        """We should index a usage if:
+        - It is a non-bandwidth usage and falls within
+          the designated time interval for data gathering.
+        - It is a bandwidth usage and its corresponding non-bandwidth usage
+          occurs within the time interval. In this situation,
+          we adjust the start date and end date of the bandwidth usage to match those of its non-bandwidth counterpart.
+          This ensures that they would be retrieved together in the report.
+        """
+        usages_should_be_indexed = []
+        for usage in usages:
+            if usage.is_bandwidth_usage():
+                cls._change_bandwidth_usage_time_interval(
+                    non_bandwidth_usage=cls._find_non_bandwidth_usage(
+                        usage_name=usage.name, usages=usages
+                    ),
+                    bandwidth_usage=usage,
+                )
+
+            if cls._is_usage_in_interval(usage=usage):
+                usages_should_be_indexed.append(usage)
+
+        return usages_should_be_indexed
+
+    @classmethod
+    def _change_bandwidth_usage_time_interval(
+        cls, non_bandwidth_usage: EquinixUsage, bandwidth_usage: EquinixUsage
+    ) -> EquinixUsage:
+        """Modifies the time of the bandwidth usage to match that of the non-bandwidth usage."""
+        bandwidth_usage.start_date = non_bandwidth_usage.start_date
+        bandwidth_usage.end_date = non_bandwidth_usage.end_date
+        return bandwidth_usage
+
+    @staticmethod
+    def _find_non_bandwidth_usage(
+        usage_name: str, usages: list[EquinixUsage]
+    ) -> EquinixUsage:
+        """Locates the non-bandwidth usage identified by the name usage_name.
+        A usage is classified as bandwidth usage if its plan includes the term "Bandwidth".
+        Presently, we recognize two types of bandwidth usages:
+          - Outbound Bandwidth
+          - Backend Transfer Bandwidth\n
+        If any such usage exists, there should be a corresponding
+        non-bandwidth usage associated with it.
```

Comment:
I am not familiar with Equinix API, am I supposed to understand what's non-bandwidth usage? If yes, please provide an example. If not, please ignore my comment :)

Reply:
non-bandwidth usages are just regular `equinix` usages e.g:

```
{
      "description": null,
      "facility": "da11",
      "metro": "da",
      "name": "ipi-ci-op-2bnzvjds-41656-1637728123456780992",
      "plan": "c3.medium.x86",
      "plan_version": "c3.medium.x86 v2 (Dell EPYC 7402P)",
      "price": 1.5,
      "quantity": 2.0,
      "total": 3.0,
      "type": "Instance",
      "unit": "hour",
      "start_date": "2023-03-20T08:30:39Z",
      "end_date": "2023-03-20T10:20:42Z"
    },
```
while its corresponding bandwidth usage will look something like -
```
{
      "description": null,
      "facility": "dc13",
      "metro": "dc",
      "name": "ipi-ci-op-2bnzvjds-41656-1637728123456780992",
      "plan": "Backend Transfer Bandwidth",
      "plan_version": "Backend Transfer Bandwidth",
      "price": 0.03,
      "quantity": 0.001739501953125,
      "total": 0.00005218505859375,
      "type": "Instance",
      "unit": "GB",
      "start_date": "2023-03-20T08:30:00Z",
      "end_date": "2023-04-20T09:30:00Z"
    },
```
same name, different plan
@rccrdpccl